File: benches/haar_wavelet.rs
use criterion::{Criterion, black_box, criterion_group, criterion_main};
use wavelers::{dwt, dwt2, idwt, idwt2, wavelet::Haar};

fn haar_wavelet_benchmark(c: &mut Criterion) {
    let signal: Vec<f64> = (0..1024).map(|x| x as f64).collect();

    c.bench_function("haar_wavelet_transform", |b| {
        b.iter(|| dwt(black_box(&signal), Haar))
    });
}

fn inverse_haar_wavelet_benchmark(c: &mut Criterion) {
    let approx: Vec<f64> = (0..512).map(|x| x as f64).collect();
    let detail: Vec<f64> = (0..512).map(|x| x as f64).collect();

    c.bench_function("inverse_haar_wavelet_transform", |b| {
        b.iter(|| idwt(black_box(&approx), black_box(&detail), Haar))
    });
}

fn haar_wavelet_benchmark_2d(c: &mut Criterion) {
    let signal: Vec<f64> = (0..1024 * 1024).map(|x| x as f64).collect();

    c.bench_function("haar_wavelet_transform_2d", |b| {
        b.iter(|| dwt2(black_box(&signal), (10, 10), Haar))
    });
}

fn inverse_haar_wavelet_benchmark_2d(c: &mut Criterion) {
    let ll: Vec<f64> = (0..512).map(|x| x as f64).collect();
    let lh: Vec<f64> = (0..512).map(|x| x as f64).collect();
    let hl: Vec<f64> = (0..512).map(|x| x as f64).collect();
    let hh: Vec<f64> = (0..512).map(|x| x as f64).collect();

    c.bench_function("inverse_haar_wavelet_transform_2d", |b| {
        b.iter(|| {
            idwt2(
                black_box(&ll),
                black_box(&lh),
                black_box(&hl),
                black_box(&hh),
                (1024, 1024),
                Haar,
            )
        })
    });
}

criterion_group!(
    benches,
    haar_wavelet_benchmark,
    inverse_haar_wavelet_benchmark,
    haar_wavelet_benchmark_2d,
    inverse_haar_wavelet_benchmark_2d,
);
criterion_main!(benches);

File: benches/python/haar_wavelet.py
import pywt
import numpy as np
import timeit

# 1D Haar Wavelet Transform
def haar_wavelet_transform(signal):
    coeffs = pywt.dwt(signal, 'haar')
    return coeffs

# Inverse 1D Haar Wavelet Transform
def inverse_haar_wavelet_transform(approx, detail):
    reconstructed_signal = pywt.idwt(approx, detail, 'haar')
    return reconstructed_signal

# 2D Haar Wavelet Transform
def haar_wavelet_transform_2d(signal):
    coeffs = pywt.dwt2(signal, 'haar')
    return coeffs

# Inverse 2D Haar Wavelet Transform
def inverse_haar_wavelet_transform_2d(ll, lh, hl, hh):
    reconstructed_signal = pywt.idwt2((ll, (lh, hl, hh)), 'haar')
    return reconstructed_signal

# Benchmarking
def benchmark():
    # Setup for 1D Haar Wavelet Transform
    setup_1d = """
import numpy as np
import pywt
signal = np.arange(1024, dtype=np.float64)
"""
    # Benchmark 1D Haar Wavelet Transform
    time_1d = timeit.timeit('haar_wavelet_transform(signal)', setup=setup_1d, globals=globals(), number=1000)
    print(f"1D Haar Wavelet Transform: {time_1d:.6f} seconds")

    # Setup for Inverse 1D Haar Wavelet Transform
    setup_inv_1d = """
import numpy as np
import pywt
approx = np.arange(512, dtype=np.float64)
detail = np.arange(512, dtype=np.float64)
"""
    # Benchmark Inverse 1D Haar Wavelet Transform
    time_inv_1d = timeit.timeit('inverse_haar_wavelet_transform(approx, detail)', setup=setup_inv_1d, globals=globals(), number=1000)
    print(f"Inverse 1D Haar Wavelet Transform: {time_inv_1d:.6f} seconds")

    # Setup for 2D Haar Wavelet Transform
    setup_2d = """
import numpy as np
import pywt
signal = np.arange(1024 * 1024, dtype=np.float64).reshape(1024, 1024)
"""
    # Benchmark 2D Haar Wavelet Transform
    time_2d = timeit.timeit('haar_wavelet_transform_2d(signal)', setup=setup_2d, globals=globals(), number=1000)
    print(f"2D Haar Wavelet Transform: {time_2d:.6f} seconds")

    # Setup for Inverse 2D Haar Wavelet Transform
    setup_inv_2d = """
import numpy as np
import pywt
ll = np.arange(512 * 512, dtype=np.float64).reshape(512, 512)
lh = np.arange(512 * 512, dtype=np.float64).reshape(512, 512)
hl = np.arange(512 * 512, dtype=np.float64).reshape(512, 512)
hh = np.arange(512 * 512, dtype=np.float64).reshape(512, 512)
"""
    # Benchmark Inverse 2D Haar Wavelet Transform
    time_inv_2d = timeit.timeit('inverse_haar_wavelet_transform_2d(ll, lh, hl, hh)', setup=setup_inv_2d, globals=globals(), number=1000)
    print(f"Inverse 2D Haar Wavelet Transform: {time_inv_2d:.6f} seconds")

if __name__ == "__main__":
    benchmark()